tests/test_subparser.py

```diff
@@ -0,0 +1,263 @@
+        assert config.action.a == 12
+        assert config.flag is False
+
+    def test_parse_positional(self):
+        config = parse(self.Config, ["Action2", "positional", "--c", "12"])
+        assert config.action.c == 12
+        assert config.action.e == "positional"
+        assert config.flag is False
+
+
+class TestDoubleAction:
+    @dataclass
+    class Config:
+        action: Union[Action1, Action2]
+        second_action: Union[Action1, Action2, Action3] = field(default_factory=lambda: Action1(a=1))
+        var: int = 12
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_is_not_allowed(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, [])
+        captured = capsys.readouterr()
+        assert "error: cannot have multiple subparser arguments" in captured.err
+
+
+class TestActionDefault:
+    @dataclass
+    class Config:
+        mode: Union[Action1, Action2, Action3] = field(default_factory=lambda: Action1(0))
+        var: int = 12
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_is_not_required(self, capsys):
+        config = parse(self.Config, [])
+        assert config.mode == Action1(0)
+        assert config.var == 12
+        assert not config.flag
+
+    def test_args(self, capsys):
+        config = parse(self.Config, ["--flag", "Action3", "--d", "0"])
+        assert isinstance(config.mode, Action3)
+        assert config.mode.d == 0
+        assert config.flag is True
+
+        # Verify that the order matters
+        with raises(SystemExit):
+            parse(self.Config, ["Action2", "--flag"])
+        captured = capsys.readouterr()
+        assert "unrecognized arguments: --flag" in captured.err
+
+
+class TestNestedAction:
+    @dataclass
+    class Config:
+        mode: Union[Action1, Action2, Action3, Action4] = field(default_factory=lambda: Action1(0))
+        var: int = 12
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_is_not_required(self, capsys):
+        config = parse(self.Config, [])
+        assert config.mode == Action1(0)
+        assert config.var == 12
+        assert not config.flag
+
+    def test_args(self, capsys):
+        config = parse(self.Config, ["Action4", "Action1", "--a", "-3"])
+        assert isinstance(config.mode, Action4)
+        assert isinstance(config.mode.sub_action, Action1)
+        assert config.mode.sub_action.a == -3
+
+
+class TestCollision:
+    @dataclass
+    class Config:
+        action: Union[Action1, Action2]
+        a: int = 2
+        d: int = 3
+
+    def test_help(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, ["Action1", "--help"], prog="prog")  # type: ignore
+        captured = capsys.readouterr()
+        assert "prog Action1 [-h] --a A [--b B]" in captured.out.replace("\n", "")
+
+    def test_collision(self, capsys):
```

Comment:
I think we expect a warning here right? should we test for that?

Reply:
I don't think we expect one (and there is none). What should we warn for?

Reply:
I thought we warned for collisions, but if not, then this is fine